src/cpu/instr.rs

```diff
@@ -40,30 +78,292 @@ struct Operation {
+                    0x7 => !(cpu_flags.contains(Flags::Z)),
+                    0x8 => cpu_flags.contains(Flags::O),
+                    0x9 => cpu_flags.contains(Flags::P),
+                    0xA => !(cpu_flags.contains(Flags::C) | cpu_flags.contains(Flags::Z)),
+                    0xB => !(cpu_flags.contains(Flags::C)), 
+                    0xC => !(cpu_flags.contains(Flags::C) | cpu_flags.contains(Flags::Z)),
+                    _ => { println!("jmp jump_code error! invalid instruction."); false },
+                };
+
+                if jump_condition_met {
+                    cpu.state.pc = enc.imm;
+                } else {
+                    cpu.state.pc = cpu.state.pc+1;
+                }
+
+            },
+            repr: |enc| {
+                let jmp_code: u8 = ((enc.rs1 << 3) as u8) +  enc.rd as u8;
+                let jmp_code_str: &str;
+
+                match jmp_code {
+                    0x0 => jmp_code_str = "jmp",
+                    0x1 => jmp_code_str = "jca",
+                    0x2 => jmp_code_str = "jeq",
+                    0x3 => jmp_code_str = "jlt",
+                    0x4 => jmp_code_str = "jgt",
+                    0x5 => jmp_code_str = "jle",
+                    0x6 => jmp_code_str = "jge",
+                    0x7 => jmp_code_str = "jne",
+                    0x8 => jmp_code_str = "jovf",
+                    0x9 => jmp_code_str = "jpar",
+                    0xA => jmp_code_str = "jgtu",
+                    0xB => jmp_code_str = "jgeu",
+                    0xC => jmp_code_str = "jleu",
+                    _ => jmp_code_str = "jump_code error",
+                }
+                format!("{} {}",jmp_code_str, enc.imm) },
+        });
+        m.insert(Opcode::CAI as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 & enc.imm as u32;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, enc.imm as u32, _out)
+            },
+            repr: |enc| format!("cai r{}, r{}", enc.rs1, enc.imm),
+        });
+        m.insert(Opcode::SAR as u8, Operation {
+            execute: |enc, cpu| {
+                let _out:u32 = (cpu.state.reg[enc.rs1 as usize] as u32 >> cpu.state.reg[enc.rs2 as usize] as u32) | (0b1111111111111111 <<  (16  - cpu.state.reg[enc.rs2 as usize]));
+                cpu.state.reg[enc.rd as usize] = _out as u16;
+                cpu.state.flags = gen_flag(false, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs2 as usize] as u32, _out);
+                cpu.state.pc = cpu.state.pc+1; 
+            },
+            repr: |enc| format!("sri r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+         });
+         m.insert(Opcode::SAI as u8, Operation {
+            execute: |enc, cpu| {
+                let _out:u32 = (cpu.state.reg[enc.rs1 as usize] as u32 >> enc.imm as u32) | (0b1111111111111111 << (16 - enc.imm)); 
```

Comment:
Arithmetic shift is complemented with ones **only** if Most Significant Bit of first operand (rs1 value) is set to one.